File: assets/lua/meta/intersect.lua
--- @meta
--- 
--- @class intersect
intersect = {}

--- 
--- @param minA math.Vector
--- @param maxA math.Vector
--- @param minB math.Vector
--- @param maxB math.Vector
--- @return enum umath::intersection::Intersect ret0
--- @overload fun(minA: vector.Vector, maxA: vector.Vector, minB: vector.Vector, maxB: vector.Vector): enum umath::intersection::Intersect
function intersect.aabb_with_aabb(minA, maxA, minB, maxB) end

--- 
--- @param originA math.Vector
--- @param rA number
--- @param originB math.Vector
--- @param rB number
--- @return bool ret0
--- @overload fun(originA: vector.Vector, rA: number, originB: vector.Vector, rB: number): bool
function intersect.sphere_with_sphere(originA, rA, originB, rB) end

--- 
--- @param start math.Vector
--- @param dir math.Vector
--- @param min math.Vector
--- @param max math.Vector
--- @param outMin any
--- @param outMax any
--- @overload fun(start: vector.Vector, dir: vector.Vector, min: vector.Vector, max: vector.Vector, outMin: any, outMax: any): 
function intersect.line_with_aabb(start, dir, min, max, outMin, outMax) end

--- 
--- @param min math.Vector
--- @param max math.Vector
--- @param origin math.Vector
--- @param r number
--- @return bool ret0
--- @overload fun(min: vector.Vector, max: vector.Vector, origin: vector.Vector, r: number): bool
function intersect.aabb_with_sphere(min, max, origin, r) end

--- 
--- @param vec math.Vector
--- @param min math.Vector
--- @param max math.Vector
--- @return bool ret0
--- @overload fun(vec: vector.Vector, min: vector.Vector, max: vector.Vector): bool
function intersect.point_in_aabb(vec, min, max) end

--- 
--- @param rayStart math.Vector
--- @param rayDir math.Vector
--- @param min math.Vector
--- @param max math.Vector
--- @return any ret0
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, min: math.Vector, max: math.Vector, precise: bool): any
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, min: math.Vector, max: math.Vector, precise: bool, pose: math.Transform): any
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, min: vector.Vector, max: vector.Vector): any
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, min: vector.Vector, max: vector.Vector, precise: bool): any
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, min: vector.Vector, max: vector.Vector, precise: bool, pose: math.Transform): any
function intersect.line_with_obb(rayStart, rayDir, min, max) end

--- 
--- @param rayStart math.Vector
--- @param rayDir math.Vector
--- @param mdl game.Model
--- @param bodyGroups 
--- @param r0 any
--- @param r1 any
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mdl: game.Model, bodyGroups: , r0: any, r1: any, precise: bool): 
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mdl: game.Model, tBodyGroups: , r0: any, r1: any, precise: bool, meshPose: math.Transform): 
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mdl: game.Model, lod: int, r0: any, r1: any): 
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mdl: game.Model, lod: int, r0: any, r1: any, precise: bool): 
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mdl: game.Model, lod: int, r0: any, r1: any, precise: bool, meshPose: math.Transform): 
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mesh: Model.Mesh, r0: any, r1: any): 
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mesh: Model.Mesh, r0: any, r1: any, precise: bool): 
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mesh: Model.Mesh, r0: any, r1: any, precise: bool, meshPose: math.Transform): 
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mesh: Model.Mesh.Sub, r0: any, r1: any): 
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mesh: Model.Mesh.Sub, r0: any, r1: any, precise: bool): 
--- @overload fun(rayStart: math.Vector, rayDir: math.Vector, mesh: Model.Mesh.Sub, r0: any, r1: any, precise: bool, meshPose: math.Transform): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mdl: game.Model, bodyGroups: , r0: any, r1: any): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mdl: game.Model, bodyGroups: , r0: any, r1: any, precise: bool): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mdl: game.Model, tBodyGroups: , r0: any, r1: any, precise: bool, meshPose: math.Transform): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mdl: game.Model, lod: int, r0: any, r1: any): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mdl: game.Model, lod: int, r0: any, r1: any, precise: bool): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mdl: game.Model, lod: int, r0: any, r1: any, precise: bool, meshPose: math.Transform): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mesh: Model.Mesh, r0: any, r1: any): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mesh: Model.Mesh, r0: any, r1: any, precise: bool): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mesh: Model.Mesh, r0: any, r1: any, precise: bool, meshPose: math.Transform): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mesh: Model.Mesh.Sub, r0: any, r1: any): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mesh: Model.Mesh.Sub, r0: any, r1: any, precise: bool): 
--- @overload fun(rayStart: vector.Vector, rayDir: vector.Vector, mesh: Model.Mesh.Sub, r0: any, r1: any, precise: bool, meshPose: math.Transform): 
function intersect.line_with_mesh(rayStart, rayDir, mdl, bodyGroups, r0, r1) end

--- 
--- @param sphereOrigin math.Vector
--- @param sphereRadius number
--- @param n math.Vector
--- @param d number
--- @return bool ret0
--- @overload fun(sphereOrigin: vector.Vector, sphereRadius: number, n: vector.Vector, d: number): bool
function intersect.sphere_with_plane(sphereOrigin, sphereRadius, n, d) end

--- 
--- @param vec math.Vector
--- @param planeTable 
--- @return bool ret0
--- @overload fun(vec: vector.Vector, planeTable: ): bool
function intersect.point_in_plane_mesh(vec, planeTable) end

--- 
--- @param origin math.Vector
--- @param dir math.Vector
--- @param n math.Vector
--- @param d number
--- @return any ret0
--- @overload fun(origin: vector.Vector, dir: vector.Vector, n: vector.Vector, d: number): any
function intersect.line_with_plane(origin, dir, n, d) end

--- 
--- @param pmin math.Vector
--- @param pmax math.Vector
--- @param pa math.Vector
--- @param pb math.Vector
--- @param pc math.Vector
--- @return bool ret0
--- @overload fun(pmin: vector.Vector, pmax: vector.Vector, pa: vector.Vector, pb: vector.Vector, pc: vector.Vector): bool
function intersect.aabb_with_triangle(pmin, pmax, pa, pb, pc) end

--- 
--- @param vec math.Vector
--- @param r number
--- @param planeTable 
--- @return int ret0
--- @overload fun(vec: vector.Vector, r: number, planeTable: ): int
function intersect.sphere_in_plane_mesh(vec, r, planeTable) end

--- 
--- @param min math.Vector
--- @param max math.Vector
--- @param planeTable 
--- @return int ret0
--- @overload fun(min: vector.Vector, max: vector.Vector, planeTable: ): int
function intersect.aabb_in_plane_mesh(min, max, planeTable) end

--- 
--- @param sphereOrigin math.Vector
--- @param radius number
--- @param coneOrigin math.Vector
--- @param coneDir math.Vector
--- @param coneAngle number
--- @return bool ret0
--- @overload fun(sphereOrigin: math.Vector, radius: number, coneOrigin: math.Vector, coneDir: math.Vector, coneAngle: number, coneSize: number): bool
--- @overload fun(sphereOrigin: vector.Vector, radius: number, coneOrigin: vector.Vector, coneDir: vector.Vector, coneAngle: number): bool
--- @overload fun(sphereOrigin: vector.Vector, radius: number, coneOrigin: vector.Vector, coneDir: vector.Vector, coneAngle: number, coneSize: number): bool
function intersect.sphere_with_cone(sphereOrigin, radius, coneOrigin, coneDir, coneAngle) end

--- 
--- @param lineOrigin math.Vector
--- @param lineDir math.Vector
--- @param v0 math.Vector
--- @param v1 math.Vector
--- @param v2 math.Vector
--- @param outT any
--- @param outUv any
--- @overload fun(lineOrigin: math.Vector, lineDir: math.Vector, v0: math.Vector, v1: math.Vector, v2: math.Vector, outT: any, outUv: any, cull: bool): 
--- @overload fun(lineOrigin: vector.Vector, lineDir: vector.Vector, v0: vector.Vector, v1: vector.Vector, v2: vector.Vector, outT: any, outUv: any): 
--- @overload fun(lineOrigin: vector.Vector, lineDir: vector.Vector, v0: vector.Vector, v1: vector.Vector, v2: vector.Vector, outT: any, outUv: any, cull: bool): 
function intersect.line_with_triangle(lineOrigin, lineDir, v0, v1, v2, outT, outUv) end

--- 
--- @param min math.Vector
--- @param max math.Vector
--- @param n math.Vector
--- @param d number
--- @return bool ret0
--- @overload fun(min: vector.Vector, max: vector.Vector, n: vector.Vector, d: number): bool
function intersect.aabb_with_plane(min, max, n, d) end

--- 
--- @param min math.Vector
--- @param max math.Vector
--- @param origin math.Vector
--- @param rot math.Quaternion
--- @param n math.Vector
--- @param d number
--- @return bool ret0
--- @overload fun(min: vector.Vector, max: vector.Vector, origin: vector.Vector, rot: math.Quaternion, n: vector.Vector, d: number): bool
function intersect.obb_with_plane(min, max, origin, rot, n, d) end

--- 
--- @param lineOrigin math.Vector
--- @param lineDir math.Vector
--- @param sphereOrigin math.Vector
--- @param sphereRadius number
--- @overload fun(lineOrigin: vector.Vector, lineDir: vector.Vector, sphereOrigin: vector.Vector, sphereRadius: number): 
function intersect.line_with_sphere(lineOrigin, lineDir, sphereOrigin, sphereRadius) end

File: core/shared/src/implementation/lua/libraries/intersect.cpp
// SPDX-FileCopyrightText: (c) 2019 Silverlan <opensource@pragma-engine.com>
// SPDX-License-Identifier: MIT
module;

module pragma.shared;

import :scripting.lua.libraries.intersect;

luabind::object Lua::intersect::line_obb(lua::State *l, const Vector3 &rayStart, const Vector3 &rayDir, const Vector3 &min, const Vector3 &max, bool precise, const pragma::math::Transform &pose)
{
	auto dist = 0.f;
	float *pDist = nullptr;
	if(precise == true)
		pDist = &dist;
	auto r = pragma::math::intersection::line_obb(rayStart, rayDir, min, max, pDist, pose.GetOrigin(), pose.GetRotation());
	if(r == false || precise == false)
		return luabind::object {l, r};
	return luabind::object {l, dist};
}

static bool get_line_mesh_result(lua::State *l, const pragma::math::intersection::LineMeshResult &res, bool precise, luabind::object &r0, luabind::object &r1)
{
	if(precise == false) {
		r0 = luabind::object {l, res.result == pragma::math::intersection::Result::Intersect};
		return false;
	}
	r0 = luabind::object {l, res.result};
	r1 = luabind::newtable(l);

	r1["hitPos"] = res.hitPos;
	r1["hitValue"] = res.hitValue;
	if(res.precise) {
		r1["triIdx"] = res.precise->triIdx;
		r1["barycentricCoords"] = Vector3 {static_cast<float>(res.precise->t), static_cast<float>(res.precise->u), static_cast<float>(res.precise->v)};
	}
	return true;
}

void Lua::intersect::line_mesh(lua::State *l, const Vector3 &rayStart, const Vector3 &rayDir, pragma::geometry::ModelSubMesh &mesh, luabind::object &r0, luabind::object &r1, bool precise, const pragma::math::Transform &meshPose)
{
	auto &origin = meshPose.GetOrigin();
	auto &rot = meshPose.GetRotation();
	pragma::math::intersection::LineMeshResult res {};
	pragma::math::intersection::line_with_mesh(rayStart, rayDir, mesh, res, precise, &origin, &rot);
	get_line_mesh_result(l, res, precise, r0, r1);
}

void Lua::intersect::line_mesh(lua::State *l, const Vector3 &rayStart, const Vector3 &rayDir, pragma::geometry::ModelMesh &mesh, luabind::object &r0, luabind::object &r1, bool precise, const pragma::math::Transform &meshPose)
{
	auto &origin = meshPose.GetOrigin();
	auto &rot = meshPose.GetRotation();
	pragma::math::intersection::LineMeshResult res {};
	pragma::math::intersection::line_with_mesh(rayStart, rayDir, mesh, res, precise, &origin, &rot);
	if(get_line_mesh_result(l, res, precise, r0, r1) == false)
		return;
	if(res.precise)
		r1["subMeshIdx"] = res.precise->subMeshIdx;
}

void Lua::intersect::line_mesh(lua::State *l, const Vector3 &rayStart, const Vector3 &rayDir, pragma::asset::Model &mdl, uint32_t lod, luabind::object &r0, luabind::object &r1, bool precise, const pragma::math::Transform &meshPose)
{
	pragma::math::intersection::LineMeshResult res {};
	pragma::math::intersection::line_with_mesh(rayStart, rayDir, mdl, res, precise, nullptr, lod, meshPose.GetOrigin(), meshPose.GetRotation());
	if(get_line_mesh_result(l, res, precise, r0, r1) == false)
		return;
	if(res.precise) {
		r1["subMeshIdx"] = res.precise->subMeshIdx;
		r1["meshGroupIdx"] = res.precise->meshGroupIndex;
		r1["meshIdx"] = res.precise->meshIdx;
	}
}

void Lua::intersect::line_mesh(lua::State *l, const Vector3 &rayStart, const Vector3 &rayDir, pragma::asset::Model &mdl, luabind::table<> tBodyGroups, luabind::object &r0, luabind::object &r1, bool precise, const pragma::math::Transform &meshPose)
{
	auto bodyGroups = Lua::table_to_vector<uint32_t>(l, tBodyGroups, 4);
	pragma::math::intersection::LineMeshResult res {};
	pragma::math::intersection::line_with_mesh(rayStart, rayDir, mdl, res, precise, &bodyGroups, 0, meshPose.GetOrigin(), meshPose.GetRotation());
	if(get_line_mesh_result(l, res, precise, r0, r1) == false)
		return;
	if(res.precise) {
		r1["subMeshIdx"] = res.precise->subMeshIdx;
		r1["meshGroupIdx"] = res.precise->meshGroupIndex;
		r1["meshIdx"] = res.precise->meshIdx;
	}
}

void Lua::intersect::line_aabb(lua::State *l, const Vector3 &start, const Vector3 &dir, const Vector3 &min, const Vector3 &max, luabind::object &outMin, luabind::object &outMax)
{
	float tMin, tMax;
	auto res = pragma::math::intersection::line_aabb(start, dir, min, max, &tMin, &tMax) == pragma::math::intersection::Result::Intersect;
	if(res == false) {
		outMin = luabind::object {l, false};
		return;
	}
	outMin = luabind::object {l, tMin};
	outMax = luabind::object {l, tMax};
}

luabind::object Lua::intersect::line_plane(lua::State *l, const Vector3 &origin, const Vector3 &dir, const Vector3 &n, float d)
{
	float t;
	bool b = pragma::math::intersection::line_plane(origin, dir, n, d, &t) == pragma::math::intersection::Result::Intersect;
	if(b == false)
		return luabind::object {l, false};
	return luabind::object {l, t};
}

bool Lua::intersect::point_in_plane_mesh(lua::State *l, const Vector3 &vec, luabind::table<> planeTable)
{
	auto planes = Lua::table_to_vector<pragma::math::Plane>(l, planeTable, 2);
	return pragma::math::intersection::point_in_plane_mesh(vec, planes.begin(), planes.end());
}

int Lua::intersect::sphere_in_plane_mesh(lua::State *l, const Vector3 &vec, float r, luabind::table<> planeTable)
{
	auto planes = Lua::table_to_vector<pragma::math::Plane>(l, planeTable, 3);
	return pragma::math::to_integral(pragma::math::intersection::sphere_in_plane_mesh(vec, r, planes.begin(), planes.end()));
}

int Lua::intersect::aabb_in_plane_mesh(lua::State *l, const Vector3 &min, const Vector3 &max, luabind::table<> planeTable)
{
	auto planes = Lua::table_to_vector<pragma::math::Plane>(l, planeTable, 3);
	return pragma::math::to_integral(pragma::math::intersection::aabb_in_plane_mesh(min, max, planes.begin(), planes.end()));
}

void Lua::intersect::line_triangle(lua::State *l, const Vector3 &lineOrigin, const Vector3 &lineDir, const Vector3 &v0, const Vector3 &v1, const Vector3 &v2, luabind::object &outT, luabind::object &outUv, bool cull)
{
	double t, u, v;
	auto bIntersect = pragma::math::intersection::line_triangle(lineOrigin, lineDir, v0, v1, v2, t, u, v, cull);
	if(bIntersect == false) {
		outT = luabind::object {l, false};
		return;
	}
	outT = luabind::object {l, t};
	outUv = luabind::object {l, ::Vector2 {u, v}};
}

luabind::object Lua::intersect::line_line(lua::State *l, const ::Vector2 &start0, const ::Vector2 &end0, const ::Vector2 &start1, const ::Vector2 &end1)
{
	auto result = pragma::math::intersection::line_line(start0, end0, start1, end1);
	if(result.has_value() == false)
		return luabind::object {l, false};
	return luabind::object {l, *result};
}
